lib/api/api_auth_webauthn.go

```diff
@@ -402,19 +373,23 @@ func (s *webauthnService) deleteOldStates() {
 	}
 }
 
-func newVolState() *WebauthnVolatileState {
-	s := WebauthnVolatileState{}
-	s.init()
+func newVolState() *apiproto.WebauthnVolatileState {
+	s := apiproto.WebauthnVolatileState{}
+	if s.Credentials == nil {
+		s.Credentials = make(map[string]*apiproto.WebauthnCredentialVolatileState, 1)
+	}
 	return &s
 }
 
-func (s *WebauthnVolatileState) init() {
-	if s.Credentials == nil {
-		s.Credentials = make(map[string]WebauthnCredentialVolatileState, 1)
-	}
+// Load volatile WebAuthn state with a read lock during loading.
+func (s *webauthnService) loadVolatileState() *apiproto.WebauthnVolatileState {
+	s.volStateMut.RLock()
+	defer s.volStateMut.RUnlock()
+	return s.loadVolatileStateRLocked()
 }
 
-func (s *webauthnService) loadVolatileState() *WebauthnVolatileState {
+// Load volatile WebAuthn state without acquiring a read lock.
+func (s *webauthnService) loadVolatileStateRLocked() *apiproto.WebauthnVolatileState {
```

Comment:
Why is a read lock needed here? The only access of member fields I see is `.miscDB`. I assumed the locking was for the maps (e.g. `registrationStates`), admittedly not basing that assumption on much. Is the locking synchronising `miscDB` access?

Reply:
Yes, it's synchronizing `miscDB` access.